File: crates/klio-parser/src/parse/stmt.rs
use super::{
    AssignOp, Block, ClassModifiers, Decl, Diagnostic, Ident, Keyword, Parser, Stmt, TokenKind,
};

impl Parser<'_, '_> {
    pub(crate) fn parse_block(&mut self) -> Option<Block> {
        let lbrace = self.expect(&TokenKind::LBrace, "`{`")?;
        let mut stmts = Vec::new();
        loop {
            self.skip_stmt_separators();
            if matches!(self.peek_kind(), TokenKind::RBrace | TokenKind::Eof) {
                break;
            }
            if let Some(s) = self.parse_stmt() {
                stmts.push(s);
            } else {
                self.recover_to_stmt_end();
            }
            // After a statement, a newline or `;` is the separator.
            if !matches!(
                self.peek_kind(),
                TokenKind::Newline | TokenKind::Semicolon | TokenKind::RBrace | TokenKind::Eof
            ) {
                let span = self.current_span();
                self.error("E0004", "expected newline or `;` between statements", span);
                self.recover_to_stmt_end();
            }
        }
        let rbrace = self.expect(&TokenKind::RBrace, "`}`")?;
        Some(Block {
            stmts,
            span: lbrace.span.join(rbrace.span),
        })
    }

    pub(crate) fn skip_stmt_separators(&mut self) {
        while matches!(self.peek_kind(), TokenKind::Newline | TokenKind::Semicolon) {
            self.pos += 1;
        }
    }

    // Single match-dispatch over statement-leading tokens; splitting would fragment it.
    #[allow(clippy::too_many_lines)]
    pub(crate) fn parse_stmt(&mut self) -> Option<Stmt> {
        let save = self.pos;
        let flags = self.skip_modifiers_with_flags();
        match self.peek_kind() {
            TokenKind::Keyword(Keyword::Val | Keyword::Var) => {
                // `val (a, b) = expr` — destructuring declaration. Falls through
                // to plain property parsing on a single name.
                let next = self.tokens.get(self.pos + 1).map(|t| &t.kind);
                if matches!(next, Some(TokenKind::LParen)) {
                    return self.parse_destructuring_decl();
                }
                self.parse_property().map(|p| Stmt::Decl(Decl::Property(p)))
            }
            TokenKind::Keyword(Keyword::Fun) => {
                let next = self.tokens.get(self.pos + 1).map(|t| &t.kind);
                if matches!(next, Some(TokenKind::Keyword(Keyword::Interface))) {
                    self.bump(); // `fun`
                    let visibility = flags.visibility;
                    let annotations = flags.annotations.clone();
                    return self
                        .parse_class(
                            ClassModifiers {
                                is_data: false,
                                is_companion: false,
                                is_enum: false,
                                is_sealed: flags.is_sealed,
                                is_open: false,
                                is_abstract: false,
                                is_inner: false,
                                is_fun_interface: true,
                                is_value: false,
                                is_annotation: false,
                                is_expect: flags.is_expect,
                                is_actual: flags.is_actual,
                            },
                            visibility,
                            annotations,
                        )
                        .map(|c| Stmt::Decl(Decl::Class(c)));
                }
                // Anonymous-function expression statement: `fun(...) ...`
                // or `fun <T>(...) ...`. No name follows the `fun` keyword.
                // For `fun <...> Ident(...)`, this is a local generic
                // function declaration — fall through to `parse_fun`.
                let after_generics = if matches!(next, Some(TokenKind::Lt)) {
                    // Skip a balanced generic param list to peek at what
                    // follows: `(` → anonymous; `Ident` → local fn.
                    let mut depth = 1i32;
                    let mut i = self.pos + 2;
                    while let Some(tok) = self.tokens.get(i) {
                        match &tok.kind {
                            TokenKind::Lt => depth += 1,
                            TokenKind::Gt => {
                                depth -= 1;
                                if depth == 0 {
                                    i += 1;
                                    break;
                                }
                            }
                            TokenKind::Eof => break,
                            _ => {}
                        }
                        i += 1;
                    }
                    self.tokens.get(i).map(|t| &t.kind)
                } else {
                    None
                };
                let is_anon = matches!(next, Some(TokenKind::LParen))
                    || matches!(after_generics, Some(TokenKind::LParen));
                if is_anon {
                    self.pos = save;
                    return self.parse_expr_or_assign_stmt();
                }
                self.parse_fun(flags).map(|f| Stmt::Decl(Decl::Function(f)))
            }
            TokenKind::Keyword(Keyword::Class | Keyword::Interface) => {
                let visibility = flags.visibility;
                let annotations = flags.annotations.clone();
                let is_value = flags.is_value || flags.is_inline;
                if flags.is_inline
                    && !flags.is_value
                    && let Some(span) = flags.inline_span
                {
                    self.diagnostics.emit(
                        Diagnostic::warning(
                            "`inline class` is deprecated; use `value class` instead",
                            span,
                        )
                        .with_code("W0001"),
                    );
                }
                self.parse_class(
                    ClassModifiers {
                        is_data: flags.is_data,
                        is_companion: false,
                        is_enum: flags.is_enum,
                        is_sealed: flags.is_sealed,
                        is_open: flags.is_open,
                        is_abstract: flags.is_abstract,
                        is_inner: flags.is_inner,
                        is_fun_interface: false,
                        is_value,
                        is_annotation: flags.is_annotation,
                        is_expect: flags.is_expect,
                        is_actual: flags.is_actual,
                    },
                    visibility,
                    annotations,
                )
                .map(|c| Stmt::Decl(Decl::Class(c)))
            }
            TokenKind::Keyword(Keyword::Object) => {
                // `object Name { … }` — local singleton. `object { … }` /
                // `object : Super { … }` is an anonymous-object *expression*
                // and falls through to expression parsing.
                let next = self.tokens.get(self.pos + 1).map(|t| &t.kind);
                if matches!(next, Some(TokenKind::Ident)) {
                    self.parse_object(
                        flags.is_data,
                        flags.is_expect,
                        flags.is_actual,
                        flags.visibility,
                    )
                    .map(|o| Stmt::Decl(Decl::Object(o)))
                } else {
                    // Roll back modifiers so the expression parser sees
                    // `object` at the head.
                    self.pos = save;
                    self.parse_expr_or_assign_stmt()
                }
            }
            TokenKind::Keyword(Keyword::Typealias) => self
                .parse_typealias(flags.visibility, flags.annotations)
                .map(|a| Stmt::Decl(Decl::TypeAlias(a))),
            _ => {
                // No declaration matched — roll back so unrelated modifiers
                // (annotations on expressions, etc.) don't get swallowed.
                self.pos = save;
                // A statement may carry leading annotations (`@Suppress(...)`
                // before a `return`/expression statement). They are runtime
                // no-ops here; consume and discard so the expression parser
                // sees the statement itself.
                let _ = self.parse_annotations();
                self.skip_nl();
                self.parse_expr_or_assign_stmt()
            }
        }
    }

    pub(crate) fn parse_destructuring_decl(&mut self) -> Option<Stmt> {
        let kw = self.bump(); // val/var
        let mutable = matches!(kw.kind, TokenKind::Keyword(Keyword::Var));
        self.expect(&TokenKind::LParen, "`(`")?;
        let mut names: Vec<Ident> = Vec::new();
        loop {
            self.skip_nl();
            if matches!(self.peek_kind(), TokenKind::RParen) {
                break;
            }
            let span = self.current_span();
            let text = self.text(span);
            // Underscore-discard binding is parsed as an ident named `_`.
            let id = if text == "_" && matches!(self.peek_kind(), TokenKind::Ident) {
                self.bump();
                Ident {
                    name: "_".into(),
                    span,
                }
            } else {
                self.parse_ident("destructured name")?
            };
            // Optional type annotation per name — consumed and discarded.
            if matches!(self.peek_kind(), TokenKind::Colon) {
                self.bump();
                let _ = self.parse_type();
            }
            names.push(id);
            self.skip_nl();
            if matches!(self.peek_kind(), TokenKind::Comma) {
                self.bump();
            } else {
                break;
            }
        }
        self.expect(&TokenKind::RParen, "`)`")?;
        self.expect(&TokenKind::Eq, "`=`")?;
        self.skip_nl();
        let init = self.parse_expr()?;
        let span = kw.span.join(init.span());
        Some(Stmt::DestructuringDecl {
            mutable,
            names,
            init,
            span,
        })
    }

    pub(crate) fn parse_expr_or_assign_stmt(&mut self) -> Option<Stmt> {
        let lhs = self.parse_expr()?;
        // Assignment forms (statement-level only).
        let op = match self.peek_kind() {
            TokenKind::Eq => Some(AssignOp::Assign),
            TokenKind::PlusEq => Some(AssignOp::Add),
            TokenKind::MinusEq => Some(AssignOp::Sub),
            TokenKind::StarEq => Some(AssignOp::Mul),
            TokenKind::SlashEq => Some(AssignOp::Div),
            TokenKind::PercentEq => Some(AssignOp::Rem),
            _ => None,
        };
        if let Some(op) = op {
            self.bump();
            self.skip_nl();
            let rhs = self.parse_expr()?;
            let span = lhs.span().join(rhs.span());
            return Some(Stmt::Assign {
                target: lhs,
                op,
                value: rhs,
                span,
            });
        }
        Some(Stmt::Expr(lhs))
    }

    // ---------- expressions (Pratt) ----------
}

File: src/parser/stmt.zig
//! Statement and block parsing: local declarations, assignments, destructuring
//! declarations, and statement-separator handling. Free functions over
//! `*Parser`.

const std = @import("std");

const ast = @import("ast");
const diagnostics = @import("diagnostics");
const lexer = @import("lexer");

const root = @import("parser.zig");
const support = @import("support.zig");
const expr = @import("expr.zig");
const file = @import("file.zig");
const class = @import("class.zig");
const members = @import("members.zig");
const types = @import("types.zig");
const control = @import("control.zig");

const Parser = root.Parser;
const ClassModifiers = root.ClassModifiers;

const AssignOp = ast.AssignOp;
const Block = ast.Block;
const Decl = ast.Decl;
const Ident = ast.Ident;
const Stmt = ast.Stmt;
const Keyword = lexer.Keyword;
const TokenKind = lexer.TokenKind;

pub fn parseBlock(p: *Parser) ?Block {
    const lbrace = support.expect(p, .LBrace, "`{`") orelse return null;
    var stmts: std.ArrayList(Stmt) = .empty;
    while (true) {
        skipStmtSeparators(p);
        switch (support.peekKind(p).*) {
            .RBrace, .Eof => break,
            else => {},
        }
        if (parseStmt(p)) |s| {
            stmts.append(p.allocator, s) catch @panic("OOM in parseBlock");
        } else {
            support.recoverToStmtEnd(p);
        }
        switch (support.peekKind(p).*) {
            .Newline, .Semicolon, .RBrace, .Eof => {},
            else => {
                const sp = support.currentSpan(p);
                support.err(p, "E0004", "expected newline or `;` between statements", sp);
                support.recoverToStmtEnd(p);
            },
        }
    }
    const rbrace = support.expect(p, .RBrace, "`}`") orelse return null;
    return Block{
        .stmts = stmts.toOwnedSlice(p.allocator) catch @panic("OOM in parseBlock"),
        .span = lbrace.span.join(rbrace.span),
    };
}

pub fn skipStmtSeparators(p: *Parser) void {
    while (switch (support.peekKind(p).*) {
        .Newline, .Semicolon => true,
        else => false,
    }) {
        p.pos += 1;
    }
}

pub fn parseStmt(p: *Parser) ?Stmt {
    const save = p.pos;
    // A label before a local DECLARATION is a runtime no-op, so the `ident@` is
    // consumed. Before an expression or loop it keeps its meaning and belongs
    // to the expression path, so strip it only when a declaration follows.
    if (std.meta.activeTag(support.peekKind(p).*) == .Ident and
        p.pos + 2 < p.tokens.len and
        (std.meta.activeTag(p.tokens[p.pos + 1].kind) == .AtNoWs or
            std.meta.activeTag(p.tokens[p.pos + 1].kind) == .AtPostWs) and
        std.meta.activeTag(p.tokens[p.pos + 2].kind) == .Keyword and
        switch (p.tokens[p.pos + 2].kind.Keyword) {
            .Val, .Var, .Fun => true,
            else => false,
        })
    {
        _ = support.bump(p); // label ident
        _ = support.bump(p); // `@`
        support.skipNl(p);
    }
    const before_mods = p.pos;
    const flags = file.skipModifiersWithFlagsLevel(p, true);
    switch (support.peekKind(p).*) {
        .LParen => {
            // `(val a, val b) = expr` carries no leading keyword; anything else
            // opening with `(` is an expression statement.
            const next: ?TokenKind = if (p.pos + 1 < p.tokens.len) p.tokens[p.pos + 1].kind else null;
            if (next != null and std.meta.activeTag(next.?) == .Keyword and (next.?.Keyword == .Val or next.?.Keyword == .Var)) {
                return parseNameBasedDestructuringStmt(p);
            }
            return parseFallthroughStmt(p, save);
        },
        .LBracket => {
            // A statement-leading `[` is only ever a positional destructuring,
            // Kotlin having no array-literal expression. But a soft-keyword
            // identifier (`data`, `value`) is eaten as a modifier above,
            // leaving its `[` index access here: when modifiers were consumed
            // this `[` is a postfix index, so parse an expression instead.
            if (p.pos != before_mods) return parseFallthroughStmt(p, save);
            return parseBracketDestructuringStmt(p);
        },
        .Keyword => |kw| switch (kw) {
            .Val, .Var => {
                // `val (a, b) = expr`; a single name falls through to plain property parsing.
                const next: ?TokenKind = if (p.pos + 1 < p.tokens.len) p.tokens[p.pos + 1].kind else null;
                if (next != null and (std.meta.activeTag(next.?) == .LParen or std.meta.activeTag(next.?) == .LBracket)) {
                    return parseDestructuringDecl(p);
                }
                const prop = members.parseLocalProperty(p, flags) orelse return null;
                const pp = p.allocator.create(ast.Property) catch @panic("OOM");
                pp.* = prop;
                return Stmt{ .Decl = Decl{ .Property = pp } };
            },
            .Fun => {
                const next: ?TokenKind = if (p.pos + 1 < p.tokens.len) p.tokens[p.pos + 1].kind else null;
                if (next != null and std.meta.activeTag(next.?) == .Keyword and next.?.Keyword == .Interface) {
                    _ = support.bump(p); // `fun`
                    const visibility = flags.visibility;
                    const annotations = flags.annotations.items;
                    const c = class.parseClass(
                        p,
                        ClassModifiers{
                            .is_data = false,
                            .is_companion = false,
                            .is_enum = false,
                            .is_sealed = flags.is_sealed,
                            .is_open = false,
                            .is_abstract = false,
                            .is_inner = false,
                            .is_fun_interface = true,
                            .is_value = false,
                            .is_annotation = false,
                            .is_expect = flags.is_expect,
                            .is_actual = flags.is_actual,
                        },
                        visibility,
                        annotations,
                    ) orelse return null;
                    return Stmt{ .Decl = Decl{ .Class = c } };
                }
                // No name after `fun`. A `fun <...> Ident(...)` is a local
                // generic declaration and falls through to `parseFun`.
                const after_generics: ?TokenKind = blk: {
                    if (next != null and std.meta.activeTag(next.?) == .Lt) {
                        // `(` means anonymous, `Ident` a local fn.
                        var depth: i32 = 1;
                        var i = p.pos + 2;
                        while (i < p.tokens.len) {
                            switch (p.tokens[i].kind) {
                                .Lt => depth += 1,
                                .Gt => {
                                    depth -= 1;
                                    if (depth == 0) {
                                        i += 1;
                                        break;
                                    }
                                },
                                .Eof => break,
                                else => {},
                            }
                            i += 1;
                        }
                        break :blk if (i < p.tokens.len) p.tokens[i].kind else null;
                    }
                    break :blk null;
                };
                const is_anon = (next != null and std.meta.activeTag(next.?) == .LParen) or
                    (after_generics != null and std.meta.activeTag(after_generics.?) == .LParen);
                if (is_anon) {
                    p.pos = save;
                    return parseExprOrAssignStmt(p);
                }
                const f = members.parseFun(p, flags) orelse return null;
                return Stmt{ .Decl = Decl{ .Function = f } };
            },
            .Class, .Interface => {
                const visibility = flags.visibility;
                const annotations = flags.annotations.items;
                const is_value = flags.is_value or flags.is_inline;
                if (flags.is_inline and !flags.is_value) {
                    if (flags.inline_span) |sp| {
                        var d = diagnostics.Diagnostic.warning(
                            "`inline class` is deprecated; use `value class` instead",
                            sp,
                        );
                        _ = d.withCode("W0001");
                        p.diagnostics.emit(p.allocator, d) catch {};
                    }
                }
                const c = class.parseClass(
                    p,
                    ClassModifiers{
                        .is_data = flags.is_data,
                        .is_companion = false,
                        .is_enum = flags.is_enum,
                        .is_sealed = flags.is_sealed,
                        .is_open = flags.is_open,
                        .is_abstract = flags.is_abstract,
                        .is_inner = flags.is_inner,
                        .is_fun_interface = false,
                        .is_value = is_value,
                        .is_annotation = flags.is_annotation,
                        .is_expect = flags.is_expect,
                        .is_actual = flags.is_actual,
                    },
                    visibility,
                    annotations,
                ) orelse return null;
                return Stmt{ .Decl = Decl{ .Class = c } };
            },
            .Object => {
                // `object Name { ... }` is a local singleton; `object { ... }`
                // and `object : Super { ... }` are EXPRESSIONS and fall through.
                const next: ?TokenKind = if (p.pos + 1 < p.tokens.len) p.tokens[p.pos + 1].kind else null;
                if (next != null and std.meta.activeTag(next.?) == .Ident) {
                    const o = class.parseObject(
                        p,
                        flags.is_data,
                        flags.is_expect,
                        flags.is_actual,
                        flags.visibility,
                        flags.annotations.items,
                    ) orelse return null;
                    return Stmt{ .Decl = Decl{ .Object = o } };
                } else {
                    p.pos = save;
                    return parseExprOrAssignStmt(p);
                }
            },
            .Typealias => {
                const a = file.parseTypealias(p, flags.visibility, flags.annotations.items) orelse return null;
                return Stmt{ .Decl = Decl{ .TypeAlias = a } };
            },
            else => return parseFallthroughStmt(p, save),
        },
        else => return parseFallthroughStmt(p, save),
    }
}

/// Rolls back so unrelated modifiers, such as annotations on expressions, are not swallowed.
fn parseFallthroughStmt(p: *Parser, save: usize) ?Stmt {
    p.pos = save;
    // A statement may carry leading annotations; they are runtime no-ops, so
    // discard them and let the expression parser see the statement.
    _ = file.parseAnnotations(p);
    support.skipNl(p);
    return parseExprOrAssignStmt(p);
}

pub fn parseDestructuringDecl(p: *Parser) ?Stmt {
    const kw = support.bump(p); // val/var
    var mutable = std.meta.activeTag(kw.kind) == .Keyword and kw.kind.Keyword == .Var;
    const bracket = std.meta.activeTag(support.peekKind(p).*) == .LBracket;
    if (!bracket) _ = support.expect(p, .LParen, "`(` or `[`") orelse return null else _ = support.bump(p);
    const entries = control.parseDestructEntries(
        p,
        if (bracket) .RBracket else .RParen,
        bracket,
        "destructured name",
    ) orelse return null;
    if (entries.any_var) mutable = true;
    _ = support.expect(p, .Eq, "`=`") orelse return null;
    support.skipNl(p);
    const init = expr.parseExpr(p) orelse return null;
    const sp = kw.span.join(init.span());
    return Stmt{ .DestructuringDecl = .{
        .mutable = mutable,
        .names = entries.names,
        .by_name = entries.by_name,
        .sources = entries.sources,
        .init = init,
        .span = sp,
    } };
}

/// Opens with `(` followed by `val`/`var`.
pub fn parseNameBasedDestructuringStmt(p: *Parser) ?Stmt {
    const open = support.bump(p); // `(`
    const entries = control.parseDestructEntries(p, .RParen, false, "destructured name") orelse return null;
    _ = support.expect(p, .Eq, "`=`") orelse return null;
    support.skipNl(p);
    const init = expr.parseExpr(p) orelse return null;
    const sp = open.span.join(init.span());
    return Stmt{ .DestructuringDecl = .{
        .mutable = entries.any_var,
        .names = entries.names,
        .by_name = entries.by_name,
        .sources = entries.sources,
        .init = init,
        .span = sp,
    } };
}

/// Opens with `[` and carries no leading keyword.
pub fn parseBracketDestructuringStmt(p: *Parser) ?Stmt {
    const open = support.bump(p); // `[`
    const entries = control.parseDestructEntries(p, .RBracket, true, "destructured name") orelse return null;
    _ = support.expect(p, .Eq, "`=`") orelse return null;
    support.skipNl(p);
    const init = expr.parseExpr(p) orelse return null;
    const sp = open.span.join(init.span());
    return Stmt{ .DestructuringDecl = .{
        .mutable = entries.any_var,
        .names = entries.names,
        .by_name = entries.by_name,
        .sources = entries.sources,
        .init = init,
        .span = sp,
    } };
}

pub fn parseExprOrAssignStmt(p: *Parser) ?Stmt {
    const lhs = expr.parseExpr(p) orelse return null;
    const op: ?AssignOp = switch (support.peekKind(p).*) {
        .Eq => AssignOp.Assign,
        .PlusEq => AssignOp.Add,
        .MinusEq => AssignOp.Sub,
        .StarEq => AssignOp.Mul,
        .SlashEq => AssignOp.Div,
        .PercentEq => AssignOp.Rem,
        else => null,
    };
    if (op) |o| {
        _ = support.bump(p);
        support.skipNl(p);
        const rhs = expr.parseExpr(p) orelse return null;
        const sp = lhs.span().join(rhs.span());
        return Stmt{ .Assign = .{
            .target = lhs,
            .op = o,
            .value = rhs,
            .span = sp,
        } };
    }
    return Stmt{ .Expr = lhs };
}
